keps/sig-apps/961-maxunavailable-for-statefulset/README.md

```diff
@@ -229,29 +429,99 @@ https://github.com/kubernetes/kubernetes/blob/v1.13.0/pkg/controller/statefulset
 ...
 ```
 
-### Risks and Mitigations
-We are proposing a new field called `maxUnavailable` whose default value will be 1. In this mode, StatefulSet will behave exactly like its current behavior.
-Its possible we introduce a bug in the implementation. The mitigation currently is that is disabled by default in Alpha phase for people to try out and give
-feedback. 
-In Beta phase when its enabled by default, people will only see issues or bugs when `maxUnavailable` is set to something greater than 1. Since people have
-tried this feature in Alpha, we would have time to fix issues.
+#### Metrics
+
+We'll add a new metric named `rolling-update-duration-seconds`, it tracks how long a statefulset takes to finish a rolling-update,
+the general logic is:
+
+when [performing update](https://github.com/kubernetes/kubernetes/blob/c984d53b31655924b87a57bfd4d8ff90aaeab9f8/pkg/controller/statefulset/stateful_set_control.go#L97-L138),
+if the `currentRevision` != `updateRevision`, then we take it as starting to rolling update, but because rolling update can not
+finish in one reconciling loop, then we may have to track it, we have two approaches:
+
+- One is add a new field in the `defaultStatefulSetControl`, like below:
+
+  ```golang
+  type defaultStatefulSetControl struct {
+    podControl        *StatefulPodControl
+    statusUpdater     StatefulSetStatusUpdaterInterface
+    controllerHistory history.Interface
+    recorder          record.EventRecorder
+
+    // <Newly Added>
+    // rollingUpdateStartTimes records when each statefulset starts to perform rolling update.
+    // key is the statefulset name, value is the start time.
+    rollingUpdateStartTimes map[string]time.Time
+  }
+  ```
+
+  Then when `currentRevision` != `updateRevision`, we'll check whether the statefulset name
+  exists in the `rollingUpdateStartTimes`, if exists, skip, if not, write the
+  start time.
```

Comment:
This doesn't work across restarts of kube-controller-manager.

Reply:
Yes, then maybe this is not a proper approach. 🤔